kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala

```diff
@@ -274,20 +255,99 @@ object ServiceDiscovery {
   }
 
   def getServerHost(zkClient: CuratorFramework, namespace: String): Option[(String, Int)] = {
+    // TODO: use last one because to avoid touching some maybe-crashed engines
+    // We need a big improvement here.
+    getServiceNodesInfo(zkClient, namespace, Some(1)) match {
+      case Seq(sn) => Some((sn.host, sn.port))
+      case _ => None
+    }
+  }
+
+  def getServiceNodesInfo(
+      zkClient: CuratorFramework,
+      namespace: String,
+      sizeOpt: Option[Int] = None): Seq[ServiceNodeInfo] = {
     try {
       val hosts = zkClient.getChildren.forPath(namespace)
-      // TODO: use last one because to avoid touching some maybe-crashed engines
-      // We need a big improvement here.
-      hosts.asScala.lastOption.map { p =>
+      val size = sizeOpt.getOrElse(hosts.size())
+      hosts.asScala.takeRight(size).map { p =>
         val path = ZKPaths.makePath(namespace, p)
-        val hostPort = new String(zkClient.getData.forPath(path), StandardCharsets.UTF_8)
-        val strings = hostPort.split(":")
+        val instance = new String(zkClient.getData.forPath(path), StandardCharsets.UTF_8)
+        val strings = instance.split(":")
         val host = strings.head
         val port = strings(1).toInt
-        (host, port)
+        val version = p.split(";").find(_.startsWith("version=")).map(_.stripPrefix("version="))
+        info(s"Get service instance:$instance and version:$version under $namespace")
+        ServiceNodeInfo(namespace, p, host, port, version)
+      }
+    } catch {
+      case e: Exception =>
+        error(s"Failed to get service node info", e)
+        Seq.empty
+    }
+  }
+
+  def createZkServiceNode(
+      conf: KyuubiConf,
+      zkClient: CuratorFramework,
+      namespace: String,
+      instance: String,
+      version: Option[String] = None,
+      external: Boolean = false): PersistentNode = {
+    val ns = ZKPaths.makePath(null, namespace)
+    try {
+      zkClient
+        .create()
+        .creatingParentsIfNeeded()
+        .withMode(PERSISTENT)
+        .forPath(ns)
+    } catch {
+      case _: NodeExistsException =>  // do nothing
+      case e: KeeperException =>
+        throw new KyuubiException(s"Failed to create namespace '$ns'", e)
+    }
+    val pathPrefix = ZKPaths.makePath(
+      namespace,
+      s"serviceUri=$instance;version=${version.getOrElse(KYUUBI_VERSION)};sequence=")
+    var serviceNode: PersistentNode = null
+    try {
+      if (external) {
+        serviceNode = new PersistentNode(
+          zkClient,
+          CreateMode.PERSISTENT_SEQUENTIAL,
+          false,
+          pathPrefix,
+          instance.getBytes(StandardCharsets.UTF_8))
+      } else {
+        serviceNode = new PersistentEphemeralNode(
```

Comment:
`PersistentEphemeralNode` is deprecated, and since we already shaded curator, it's safe to drop it. We can do it in the following PRs.

Reply:
ok